crates/net/downloaders/src/headers/concurrent.rs

```diff
@@ -0,0 +1,599 @@
+        }
+
+        if retries < self.request_retries {
+            // re-submit the request
+            self.submit_request(request, retries + 1);
+            None
+        } else {
+            tracing::trace!(
+                target: "downloaders::headers",
+                "Ran out of retries"
+            );
+            Some(error)
+        }
+    }
+
+    /// Attempts to validate the buffered responses
+    ///
+    /// Returns an error if the next expected response was popped, but failed validation
+    fn try_validate_buffered(&mut self) -> Option<HeadersResponseError> {
+        loop {
+            // Check to see if we've already received the next value
+            let next_response = self.buffered_responses.peek_mut()?;
+            if next_response.block_number() == self.next_request_block_number {
+                let OrderedHeadersResponse { headers, request, retries, peer_id } =
+                    PeekMut::pop(next_response);
+                if let Err(err) = self.process_next_headers(request, headers, peer_id, retries) {
+                    return Some(err)
+                }
+            } else {
+                return None
+            }
+        }
+    }
+
+    /// Starts a request future
+    fn submit_request(&mut self, request: HeadersRequest, retries: usize) {
+        let client = Arc::clone(&self.client);
+        let fut = HeadersRequestFuture {
+            request: Some(request.clone()),
+            fut: Box::pin(async move { client.get_headers(request).await }),
+            retries,
+        };
+        self.in_progress_queue.push(fut);
+    }
+
+    /// Clears all requests/responses.
+    fn clear(&mut self) {
+        self.queued_validated_headers.clear();
+        self.buffered_responses.clear();
+        self.in_progress_queue.clear();
+    }
+
+    fn terminate(&mut self) {
+        self.is_terminated = true;
+        self.clear();
+    }
+}
+
+impl<C, H> Downloader for ConcurrentDownloader<C, H>
+where
+    C: Consensus,
+    H: HeadersClient,
+{
+    type Client = H;
+    type Consensus = C;
+
+    fn client(&self) -> &Self::Client {
+        self.client.borrow()
+    }
+
+    fn consensus(&self) -> &Self::Consensus {
+        self.consensus.borrow()
+    }
+}
+
+impl<C, H> HeaderDownloader2 for ConcurrentDownloader<C, H>
+where
+    C: Consensus,
+    H: HeadersClient,
+{
+}
+
+impl<C, H> Stream for ConcurrentDownloader<C, H>
+where
+    C: Consensus + 'static,
+    H: HeadersClient + 'static,
+{
+    type Item = DownloadResult<Vec<SealedHeader>>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.get_mut();
+
+        if this.is_terminated {
+            return Poll::Ready(None)
+        }
+
+        // yield next batch
+        if this.queued_validated_headers.len() > this.stream_batch_size {
+            let next_batch = this.queued_validated_headers.split_off(this.stream_batch_size);
```

Comment:
`next_batch` will be at range `[this.stream_batch_size, this.queued_validated_headers.len())`. not sure this is intentional. if it is, why would we want to keep `[0, this.stream_batch_size)` buffered instead of yielding them?